### PR TITLE
HW08: Nikita Dukin

### PR DESCRIPTION
**Pull request by Fant1k34**

### CODE REVIEW

File: test/Property.hs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+generateRandomOperation opMode = case opMode of
+  ALL -> (do
+    randomValue <- Gen.integral_ (Range.constant 0 4)
+
+    return (case randomValue of
+      0 -> Plus
+      1 -> Min
+      2 -> Div
+      3 -> Mul
+      _ -> In
```
Comment (kajigor) [inline test/Property.hs:89]: You can use `Gen.element` for generating one of the predefined set of values.

File: test/Property.hs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+        return (CE expr1 randomOperation expr2)
+        )
+
+
+propCorrectness :: Property
+propCorrectness = property (do
+    xs <- forAll $ Gen.list (Range.linear 0 100) Gen.alpha
+    xs === xs
+    )
```
Comment (kajigor) [inline test/Property.hs:155]: What is this property supposed to check?

File: test/Property.hs [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+
+propEvalExpressionWithSqrtError :: Property
+propEvalExpressionWithSqrtError = property (do
+  expr0 <- forAll $ generateExprSample 13 NEGATIVE_VALUE SAFE True
+
+  case evaluate (fmap fromInteger expr0) [] of
+    Left (OutOfPossibleValuesError _ _) -> assert True
+    value -> show False === show value ++ " got from " ++ show expr0
```
Comment (kajigor) [inline test/Property.hs:164]: There is an easier way to fail a test: see `failure`

File: test/Property.hs [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    Left comment -> show False === show comment ++ " got from " ++ show expr
+    Right resutOfExpr ->
+      case evaluate (fmap fromInteger subExpr1) [] of
+        Left comment -> show False === show comment ++ " got from " ++ show subExpr1
+        Right subValue1 ->
+          case evaluate (fmap fromInteger subExpr2) [] of
+            Left comment -> show False === show comment ++ " got from " ++ show subExpr2
+            Right subValue2 -> resutOfExpr === subValue1 ** subValue2
+  )
+
```
Comment (kajigor) [inline test/Property.hs:264]: Reduce code duplication by parameterizing this over the binary operation

File: test/Property.hs [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+propEvalExpressionWithVar :: Property
+propEvalExpressionWithVar = property (do
+  exprVar <- forAll $ generateExprSample 3 VAR_X SAFE False
+  exprAnySafe <- forAll $ generateExprSample 7 ALL_POSSIBLE_VALUES SAFE False
+  randomOperation <- forAll $ generateRandomOperation SAFE
+
+  let expr = CE exprAnySafe randomOperation exprVar
+
+  case evaluate (fmap fromInteger expr) [("x", 123)] of
```
Comment (kajigor) [inline test/Property.hs:214]: You can also generate the var mapping :)

**Review by kajigor — CHANGES_REQUESTED** [review]

**Comment by kajigor** [comment]
You did well. 👍  There are some aspects of how you write your code that you can improve.

File: test/Property.hs (hunk 1 above)
Comment (Fant1k34) [reply test/Property.hs:89]: Thank you for comment
It worked well and made code look beauty ;)

File: test/Property.hs (hunk 3 above)
Comment (Fant1k34) [reply test/Property.hs:164]: Done, thanks

File: test/Property.hs (hunk 4 above)
Comment (Fant1k34) [reply test/Property.hs]: Thanks for comment, fixed it
I introduced several arguments to wrap subExpr, apply binary operation

File: test/Property.hs (hunk 5 above)
Comment (Fant1k34) [reply test/Property.hs:163]: Yes, i can ahahaahha
But may be it would be enough writing tests for this time?))

**Comment by kajigor** [comment]
ok

**Pull request closed without merging**